Search.py

```diff
@@ -236,6 +236,64 @@ def can_beat_game(self, scan_for_items=True):
+                for i in goal.items:
+                    # Every world has the same set of goals, but reachable item quantity varies per-world in beatable only
+                    # Avert your eyes, here be spaghetti
+                    beaten = (beaten and all(map(State.has, self.state_list, itertools.repeat(i['name']), [state.world.goal_categories[category_name].get_goal(goal.name).get_item(i['name'])['quantity'] for state in self.state_list])))
+            if (goal.locations is not None):
+                for l in goal.locations:
+                    beaten = (beaten and all(map(self.can_reach_spot, self.state_list, itertools.repeat(l['name']), itertools.repeat(l['age']))))
+
+            if beaten:
+                valid_goals.append(goal.name)
+
+        if scan_for_items:
+            # collect all available items
+            # make a new search since we might be iterating over one already
+            search = self.copy()
+            search.collect_locations()
+            for goal in test_goals:
+                beaten = True
+                # if every state beat the goal, then return True
+                if (goal.items is not None):
+                    for i in goal.items:
+                        beaten = (beaten and all(map(State.has, search.state_list, itertools.repeat(i['name']), [state.world.goal_categories[category_name].get_goal(goal.name).get_item(i['name'])['quantity'] for state in self.state_list])))
+                if (goal.locations is not None):
+                    for l in goal.locations:
+                        beaten = (beaten and all(map(search.can_reach_spot, search.state_list, itertools.repeat(l['name']), itertools.repeat(l['age']))))
+                if beaten and not goal.name in valid_goals:
+                    valid_goals.append(goal.name)
+        
+        return valid_goals
+
+
+    def update_reachable_goals(self, category_name):
+        # Only reduce goal item quantity if minimum goal requirements are reachable, 
+        # but not the full goal quantity. Primary use is to identify reachable
+        # skull tokens, triforce pieces, and plentiful item duplicates with
+        # All Locations Reachable off. Beatable check is required to prevent
+        # inadvertently setting goal requirements to 0 or some quantity less
+        # than the minimum
+        # Relies on consistent list order to reference between current state
+        # and the copied search
+        search = self.copy()
+        search.collect_locations()
+        for index, state in enumerate(self.state_list):
+            for goal in state.world.goal_categories[category_name].goals:
+                if (goal.items is not None):
+                    if all(map(State.has_item_goal, itertools.repeat(search.state_list[index]), goal.items)):
+                        for i in goal.items:
+                            i['quantity'] = min(search.state_list[index].item_count(i['name']), i['quantity'])
```

Comment:
pro-tip: you can use the bound member function directly in map :)
```suggestion
                if goal.items:
                    if all(map(state.has_item_goal, goal.items)):
                        for i in goal.items:
                            i['quantity'] = min(state.item_count(i['name']), i['quantity'])
```